### PR TITLE
Add Jennifer Recipe

### PR DESCRIPTION
**Pull request by faustinoaq**
Hi @imdrasil @damianham 

This PR tries to add a Jennifer Recipe to replace https://github.com/amberframework/amber/pull/766

This will be easier to maintain, contribute, test and deploy because amber templates a becoming too complex, so recipes to rescue!!! :wink: 

This is an initial implementation, so, there are some issues to fix:

1. Recipes doesn't support migrations or auth yet (only app, controllers, models, and scaffold): see: https://github.com/amberframework/amber/blob/master/src/amber/cli/recipes/recipe.cr#L25
2. Jennifer (and some of its dependencies) has wrong tags: see https://github.com/imdrasil/jennifer.cr/issues/137 (this causes broken dependencies, see: https://github.com/crystal-lang/shards/issues/183)
3.  `amber new` doesn't allow other models, just granite and crecto, see: https://github.com/amberframework/amber/issues/827
4. Jennifer doesn't support SQLite yet, see: https://github.com/imdrasil/jennifer.cr#important-restrictions

Roadmap

- [x] Add `jennifer_spec.cr` and auto-deploy support using Travis
- [x] Add Jennifer recipe README
- [x] Add main app folder with Jennifer dependencies and `sam.cr` file
- [x] Add models using Jennifer and model specs
- [x] Complete scaffolding (views and controllers) using Jennifer and scaffolding specs WIP :warning: 
- [ ] Add auth using Jennifer and auth specs
- [x] Fix formatting and other specs, see #15

### CODE REVIEW

**Comment by imdrasil** [comment]
@faustinoaq after merging master into this branch it is impossible to find what are the original changes. Would you mind if I reset last merge commit and do rebase instead of merge (maybe I will have to squash commits)? The opposite solution is to create separate PR with my original changes and add changes what are done here. What do you think?

**Comment by faustinoaq** [comment]
@imdrasil Oh, my bad, I just messed up the code, I tried to fix it unsuccessfully, can you create another PR with Jennifer changes? :sweat_smile:

**Comment by faustinoaq** [comment]
@imdrasil I'm just wondering why #4 didn't have this merging issue, @damianham WDYT? :sweat_smile:

**Comment by faustinoaq** [comment]
@imdrasil I just fixed merge issues, All Jennifer code has been synced with master :sparkles: :tada:

File: jennifer/model/src/models/{{name}}.cr.lqd [hunk 1]
```diff
@@ -0,0 +1,32 @@
+{% assign id_field_type = "Primary32" -%}
```
Comment (imdrasil) [inline jennifer/model/src/models/{{name}}.cr.lqd:1]: ATM Jennifer by default use Int32 field database type for a primary key.

File: jennifer/model/src/models/{{name}}.cr.lqd [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+{% endfor -%}
+class {{ class_name }} < Jennifer::Model::Base
+  with_timestamps
+
+  mapping(
+    id: {{ id_field_type }},
+{%- for field in fields -%}
+{%- if field.name != "id" -%}
+{%- if field.type == "reference" %}
+    {{ field.name }}_id: { type: Int32? },
```
Comment (imdrasil) [inline jennifer/model/src/models/{{name}}.cr.lqd]: Same with foreign keys ^^^

File: jennifer/scaffold/controller/src/controllers/{{name}}_controller.cr.lqd [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    if {{ name }} = {{ class_name }}.find(params["id"])
+      {{ name }}.destroy
+    else
+      flash["warning"] = "{{ class_name }} with ID #{params["id"]} Not Found"
+    end
+    redirect_to "/{{ name | pluralize }}"
+  end
+
+  def {{ name }}_params
+    {{ class_name }}.build_params(
```
Comment (imdrasil) [inline jennifer/scaffold/controller/src/controllers/{{name}}_controller.cr.lqd]: Here I need some advice: Jennifer has own built-in validation + she can't use `Hash(String, String)` as an argument for building objects. Should I use arguments validation anyway?

**Comment by faustinoaq** [comment]
Hi @imdrasil Thank you for your latest commits, very appreciated! :+1: 

Just a comment:

The `{{"  "}}` code is required to preserve white space because liquid templates removes all blank spaces when `{%- -%}` is used.

Current specs are failing because formatting issues on generated code:

```
Error: formatting 'src/controllers/category_controller.cr' produced changes
Error: formatting 'src/controllers/product_controller.cr' produced changes
Error: formatting 'src/controllers/comment_controller.cr' produced changes
Error: formatting 'src/models/category.cr' produced changes
Error: formatting 'src/models/product.cr' produced changes
Error: formatting 'src/models/comment.cr' produced changes
Error: 'config/initializers/database.cr' has syntax errors
Error: formatting 'spec/controllers/category_controller_spec.cr' produced changes
Error: formatting 'spec/controllers/product_controller_spec.cr' produced changes
Error: formatting 'spec/controllers/comment_controller_spec.cr' produced changes
Error: formatting 'spec/models/category_spec.cr' produced changes
Error: formatting 'spec/models/product_spec.cr' produced changes
Error: formatting 'spec/models/comment_spec.cr' produced changes
```

^^ These don't include `.ecr` nor `.slang` generated templates, so those should be checked manually :sweat_smile:

File: jennifer/scaffold/controller/src/controllers/{{name}}_controller.cr.lqd (hunk 3 above)
Comment (damianham) [reply jennifer/scaffold/controller/src/controllers/{{name}}_controller.cr.lqd]: The validation in the controller is validation of the incoming parameters before it hits the ORM layer so I think you would leave that method as it is.

**Comment by eliasjpr** [comment]
Ping! @imdrasil @damianham

**Comment by imdrasil** [comment]
Hi. I had some stuff to do. Will return to this from Sunday.

**Comment by eliasjpr** [comment]
@imdrasil awesome I saw the new version of Jennifer and it would be great to have this recipe available for the next release

File: basic/crecto/app/config/environments/development.yml.lqd [hunk 4]
```diff
@@ -7,8 +7,7 @@ logging:
   filter:
     - password
     - confirm_password
-  skip:
-    -
+  skip: []
```
Comment (imdrasil) [inline basic/crecto/app/config/environments/development.yml.lqd:10]: such kind of change is made because the previous case created `[nil]` array.

**Comment by imdrasil** [comment]
@faustinoaq , @eliasjpr please take a look at my changes when have some free time

**Comment by eliasjpr** [comment]
This is awesome @imdrasil. Can you add a readme to the recipe on how to setup. I think it will come in handy.

File: spec/jennifer_spec.cr [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    build_app.should be_true
+
+    runner.file("src/models/product.cr").should eq(runner.fixtures("jennifer/models/product.cr"))
+    runner.file("src/controllers/product_controller.cr").should eq(runner.fixtures("jennifer/controllers/products_controller_ecr.cr"))
+
+    test_app.should be_true
+  ensure
+    runner.cleanup
+  end
+end
```
Comment (eliasjpr) [inline spec/jennifer_spec.cr]: New line missing

File: jennifer/app/src/assets/stylesheets/main.scss [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+.nav-item-auth {
+  &.nav-item-auth-signout, &.nav-item-auth-signup {
+    order: 9999;
+  }
+
+  &.nav-item-auth-profile, &.nav-item-auth-signin {
+    order: 9990;
+    margin-left: auto;
+  }
+}
```
Comment (eliasjpr) [inline jennifer/app/src/assets/stylesheets/main.scss]: New Line

File: jennifer/app/src/views/home/index.ecr.lqd [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+    <h2><%= t "welcome_to_amber" %></h2>
+    <p>Thank you for trying out the Amber Framework.  We are working hard to provide a super fast and reliable framework that provides all the productivity tools you are used to without sacrificing the speed.</p>
+    <div class="list-group">
+      <a class="list-group-item list-group-item-action" target="_blank" href="https://docs.amberframework.org">Getting Started with Amber Framework</a>
+      <a class="list-group-item list-group-item-action" target="_blank" href="https://github.com/imdrasil/jennifer.cr/blob/master/docs/index.md">Getting Started with Jennifer ORM</a>
+      <a class="list-group-item list-group-item-action" target="_blank" href="https://github.com/veelenga/awesome-crystal">List of Awesome Crystal projects and shards</a>
+      <a class="list-group-item list-group-item-action" target="_blank" href="https://crystalshards.xyz">What's hot in Crystal right now</a>
+    </div>
+  </div>
+</div>
```
Comment (eliasjpr) [inline jennifer/app/src/views/home/index.ecr.lqd]: New Line

Comment (imdrasil) [reply jennifer/app/src/views/home/index.ecr.lqd]: Should I add a newline in similar places for other recipes?

Comment (eliasjpr) [reply jennifer/app/src/views/home/index.ecr.lqd]: We should do that in a separate PR

**Review by eliasjpr — APPROVED** [review]

**Comment by imdrasil** [comment]
@faustinoaq please take a look at the changes when have some free time

**Comment by faustinoaq** [comment]
@imdrasil Yeah, sorry for late reply, I'm organizing a lot of things first. I'll check this, Don't worry :wink:

**Pull request closed without merging**